pkg/rdt/rdt.go

```diff
@@ -258,64 +274,93 @@ func (c *control) configureResctrl(conf config) error {
 	c.DebugBlock("", "applying resolved config: |\n%s", utils.DumpJSON(conf))
 
 	// Remove stale resctrl groups
-	existingClasses, err := c.classesFromResctrlFs()
+	classesFromFs, err := c.classesFromResctrlFs()
 	if err != nil {
 		return err
 	}
 
-	for _, cls := range existingClasses {
-		if _, ok := conf.Classes[cls.name]; !ok {
+	for name, cls := range classesFromFs {
+		if _, ok := conf.Classes[cls.name]; cls.name != RootClassName && !ok {
 			tasks, err := cls.GetPids()
 			if err != nil {
 				return rdtError("failed to get resctrl group tasks: %v", err)
 			}
 			if len(tasks) > 0 {
 				return rdtError("refusing to remove non-empty resctrl group %q", cls.relPath(""))
 			}
-			err = os.Remove(cls.path(""))
+			log.Debug("removing existing resctrl group %q", cls.relPath(""))
+			err = groupRemoveFunc(cls.path(""))
 			if err != nil {
 				return rdtError("failed to remove resctrl group %q: %v", cls.relPath(""), err)
 			}
+
+			delete(c.classes, name)
 		}
 	}
 
-	// Start with fresh set of classes. Root class is always present
-	c.classes = make(map[string]*ctrlGroup, len(conf.Classes))
-	c.classes[RootClassName], err = newCtrlGroup(c.resctrlGroupPrefix, RootClassName)
-	if err != nil {
-		return err
+	for name, cls := range c.classes {
+		if _, ok := conf.Classes[cls.name]; cls.name != RootClassName && !ok {
+			log.Debug("dropping stale class %q", name)
+			delete(c.classes, name)
+		}
+	}
+
+	if _, ok := c.classes[RootClassName]; !ok {
+		log.Panic("root class missing from runtime data, re-adding...")
```

Comment:
typo/thinko: should be `log.Warning()` or `log.Error()` since you try to continue...

Reply:
Is it? I think of panic like kernel panic. Something that really should not happen but we're able to continue. Maybe this is a philosophical question 😄  I think fatal is the kind of error where you bail out immediately. Or then return an error. But in this case we're able to recover from the problem

Reply:
Maybe I miss something here, but I don't understand why it would be philosophical instead of purely practical ? What I meant is that there is no point in placing code in a basic block after a call to `log.Panic()` or `log.Fatal()`. The former `panic()`s, the latter calls `os.Exit(1)`. In both cases the remaining code inside the basic block is unreachable.

Reply:
Ah okay, I guess with 'philosophical question' you meant the semantics, IOW whether `log.Panic()` should effectively exit or not. I think `panic()'ing`/exiting is the right to do because that's what I believe most people will expect based on the (semantic naming) reference to both golang `panic()` and kernel panics. If it was called `log.Oops()` then I'd understand the expectation of continued execution ;-)

Reply:
Yeah, you're absolutely correct 😪  I got fooled by some of my logging experiments where the logger was wired differently. And, true, this wouldn't work even if `recover` was called. Thanks for spotting this and correcting me. Changed now to `Warn()`